### PR TITLE
refactor(draft-editor): separate draft editor from lilith-core

### PR DESCRIPTION
**Pull request by caesarWHLee**
# Notable Change
- lilith-draft-editor: new package lilith-draft-editor to manage all draft.js' related logic
- lilith-core: remove draft.js' related logic and let user(lilith-mirrormedia...) dynamically choose the draftEditor for the website
- lilith-mirrormedia/lilith-readr: call get draftEditor with their own website name

## lilith-draft-editor
The file structure inside /src:
```
/draft-js    (shared draft logic for all website)
    /block-renderer    (the render blocks draft.js `blockRendererFn` use, separate from buttons)
    /buttons    (where our custom draft.js buttons stays) 
    /editor    (all the other files)
    /entity-decorator    (the decoratos draft.js Editor use, separate from buttons)
/website
    /mirrormedia
        /custom    
            /block-renderer    (custom block to render different UI or handle different data structure in mirrormedia's CMS and Next.js)
            /entity-decorator    (custom decorator)
            /selector    (custom selector to modify gql query in post/photo/video query)
        /draft-editor (provide DraftEditor for CMS)
        /draftrenderer (provide DraftRenderer fro Next.js)
    /readr (same structure like ../mirrormedia)
```
This package maintains a group of buttons and the websites choose some of them to use. To render the content created from buttons, there will be **block renderers** and **entity-decorators**. 

Since every website's CMS has different GQL schema for a list and Next.js has different UI for each content added by a button. There will be multiple block-renderers and entity-decorators according to how many websites using these package.

For a website like mirrormedia, the render block should be slightly different for CMS and Next.js because some render block has ability to edit in CMS after created. 

these buttons are listed below:
- InfoboxBlock
- TableBlock
- ColorBoxBlock
- BGImageBlock
- BGVideoBlock
- SideIndexBlock

For these buttons, there will be two versions of blocks: one for CMS and one for Next.js.
The convention is naming with 'EditorBlock' or 'Block', like InfoBoxEditorBlock for CMS and InfoBoxBlock for Next.js.

### For CMS to use (DraftEditor)
1. Check if the website exist in lilith-darft-editor/src/website. If not, make a copy from lilith-draft-editor/src/website/mirrormedia.
    - change the export lilith-draft-editor/src/website/{website}/index.js from `MirrorMedia` to the website name with first letter captitalized.
3. In lilith-core, check if the website exist in lilith-core/src/custom-fields/rich-text-editor/views. If not, make a copy from mirrormedia folder in the same location.
    - make sure to `import {Website} from '@mirrormedia/lilith-draft-editor'`
4. For lilith-${website}, make sure add property `website` to the calling of `customFields.richTextEditor`
    ex:
    ```
    //before
    content: customFields.richTextEditor({
      label: '敘述',
    }),
    //after
    content: customFields.richTextEditor({
      label: '敘述',
      website: 'mirrormedia',
    }),
    ```
5. Modify the draft editor by the spec of specific website
    - used buttons in CMS
    - gql for Post/Photo/Video and their blocks (RelatedPostBlock, ImageBlock, SlideshowBlock, BGImageBlock, BGVideoBlock..)
    - usually the render block will be implemented by the owner developer Next.js for the website, just make sure every thing looks ok in CMS
6. Run the CMS and test if everything works fine

### For Next.js to use (DraftRenderer)

* Make sure the steps are done for CMS to use lilith-draft-editor

#### To modify content added from buttons

- atomic block: modify component in draft-editor/src/website/${website}/custom/block-renderer 
- decorator: modify component in draft-editor/src/website/${website}/custom/entity-decorator
- built-in block: write css in draft-editor/src/website/${website}/draft-renderer/index.ts
    ```
      /* Draft built-in buttons' style */
      .public-DraftStyleDefault-header-two {
      }
      .public-DraftStyleDefault-header-three {
      }
      .public-DraftStyleDefault-header-four {
      }
      .public-DraftStyleDefault-blockquote {
      }
      .public-DraftStyleDefault-ul {
      }
      .public-DraftStyleDefault-unorderedListItem {
      }
      .public-DraftStyleDefault-ol {
      }
      .public-DraftStyleDefault-orderedListItem {
      }
      /* code-block */
      .public-DraftStyleDefault-pre {
      }

    ```

#### To use DraftRenderer

```
  import { MirrorMedia } from '@mirrormedia/lilith-draft-editor'
  const {RichTextEditor} = MirrorMedia.DraftRenderer
  // use it with the raw `rawContentBlock` fetched from gql
  <RichTextEditor rawContentBlock={postContent}/>
```

### CODE REVIEW

**Comment by dyfu95** [comment]
Question: 
As @nickhsine mentioned before, maybe `mirrormedia` should rename to `mirrorweekly` ?

**Comment by dyfu95** [comment]
Question: 
So if I modify the atomic-block style in `draft-editor/src/website/${website}/custom/block-renderer`, it will change the style of which in CMS and Next.js at the same time ?

**Comment by dyfu95** [comment]
Question: 
Since `mirrorweekly` has to change style of atomic block on different article type, is it possible to add class name on every atomic block, so frontend can modify style at Nextjs website ?

**Comment by caesarWHLee** [comment]
> Question: As @nickhsine mentioned before, maybe `mirrormedia` should rename to `mirrorweekly` ?

Yes, for now in draft-editor the reason I use `mirrormedia` for the name of the website is to maintain consistancy to the lilith-mirrormedia. Having another PR to modify all the naming from `mirrormedia` to `mirrorweekly` will be a necessary.

> Question:
So if I modify the atomic-block style in `draft-editor/src/website/${website}/custom/block-renderer`, it will change the style of which in CMS and Next.js at the same time ?

Yes, the render block (both block-renderer and entity-decorator inside `website/${website}/custom` is designed to shared to both DraftEditor (for CMS) and DraftRenderer (for Next.js) which achieve the WYSIWYG. 
** note: Some block-renderers there are two components exported like `InfoBoxBlock` and `InfoBoxEditorBlock`, the latter is created to give ability for the component to edit after created from the button. Make sure not to break the dependency that `InfoBoxEditorBlock` use `InfoBoxBlock` to render the content if `InfoBoxBlock` is modified.

> Question:
Since `mirrorweekly` has to change style of atomic block on different article type, is it possible to add class name on every atomic block, so frontend can modify style at Nextjs website ?

Sure, fell free to add class name. The block-renderer in the custom folder should be taken as a component used directly to Next.js but it stays inside the draft-editor to encapsulate the render logic close to the place we use draft-js `Editor`.

**Comment by nickhsine** [comment]
Question:
前端 weekly-next 使用的 RichTextEditor 的版本會 import buttons 嗎？

File: packages/draft-editor/package.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    "draft-js": "^0.11.7"
+  },
+  "peerDependencies": {
+    "react": "18.1.0",
+    "react-dom": "^18.1.0",
+    "styled-components": "5.3.5",
+    "@keystone-6/core": "1.1.1",
+    "@keystone-ui/button": "^6.0.0",
+    "@keystone-ui/fields": "^6.0.0",
+    "@keystone-ui/modals": "^5.0.0"
```
Comment (nickhsine) [inline packages/draft-editor/package.json:34]: Question:
前端 weekly-next 使用的 draft editor 應該是沒有 buttons 的版本，
如此一來，weekly-next 會需要安裝 @keystone-6 和 @keystone-ui 相關的 pkgs 嗎？

Comment (caesarWHLee) [reply packages/draft-editor/package.json:34]: > Question:
前端 weekly-next 使用的 RichTextEditor 的版本會 import buttons 嗎？

對，weekly-next不會用到buttons

weekly-next 使用的draft renderer不會需要 @keystone-6 和 @keystone-ui 相關的pkgs，
不過因為目前是把draft renderer和draft editor放在同一個package，所以才會寫在peer dependency

**Comment by caesarWHLee** [comment]
基於之前draft-editor和draft-renderer在同一個package的情況下，next.js在import lilith-draft-editor的時候會把CMS中才會用到的buttons載入，而buttons之中又有用到@keyston-ui相關的套件，導致在使用上next.js需要同時安裝一堆不需要的套件 (更別提在next.js上安裝@keystone-core會直接無法正常運作的問題..)，所以在新的commit中將CMS和Next.js使用的draft package區分開來

# draft-editor (for CMS)

架構保留原本的結構，移除各個website中draft-renderer資料夾以及大部分的custom資料夾中的block-renderer，block-renderer統一採用從draft-renderer中import進來的components。

# draft-renderer  (for Next.js)

架構同原本的draft-editor，移除各個website中draft-editor資料夾。

File: packages/draft-renderer/package.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "dev": "make dev",
+    "build": "make build",
+    "clean": "make clean",
+    "build-lib": "make build-lib"
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/readr-media/react.git",
```
Comment (nickhsine) [inline packages/draft-renderer/package.json:15]: nitpick: 
`url` 的網址好像錯了？

File: packages/draft-renderer/package.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    "readr",
+    "draft.js"
+  ],
+  "license": "MIT",
+  "dependencies": {
+    "draft-js": "^0.11.7"
+  },
+  "peerDependencies": {
+    "react": "18.1.0",
+    "react-dom": "^18.1.0",
```
Comment (nickhsine) [inline packages/draft-renderer/package.json:29]: nitpick:
看要不要和 `react` 一樣鎖死在 `18.1.0`？

**Review by nickhsine — COMMENTED** [review]
Question 🙋 :
1. `draft-editor/src/draft-js/block-renderer` 底下的檔案是設計成 default renderer，但可以被 `draft-editor/src/website/mirrormedia/custom/block-renderer` 底下的檔案給 overwrite 嗎？
2. 一樣的問題也想問 `draft-renderer`，因為其資料夾結構與 `draft-editor` 一樣。
3. 好奇 `draft-renderer/src/draft-js/block-renderer` 是不是不小心加入的？因為我想說 `draft-renderer` 是不是不需要 default block renderer，可以直接維護 `website` 底下每個 project 各自的 block renderer？
4. 好奇 `draft-editor/src/draft-js/block-renderer` 是不是也是不小心加入的？好像也可以直接使用 `website` 底下的 block renderer 就好？

File: packages/draft-renderer/package.json (hunk 3 above)
Comment (caesarWHLee) [reply packages/draft-renderer/package.json]: 感謝建議，已修正在[ba73ee4](https://github.com/mirror-media/Lilith/pull/213/commits/ba73ee4bf704294f61e236d0d55db1f347915b60)

File: packages/draft-renderer/package.json (hunk 2 above)
Comment (caesarWHLee) [reply packages/draft-renderer/package.json]: 感謝debug，已修正在[ba73ee4](https://github.com/mirror-media/Lilith/pull/213/commits/ba73ee4bf704294f61e236d0d55db1f347915b60)

File: packages/draft-editor/package.json [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    "readr",
+    "draft.js"
+  ],
+  "license": "MIT",
+  "dependencies": {
+    "draft-js": "^0.11.7"
+  },
+  "peerDependencies": {
+    "react": "18.1.0",
+    "react-dom": "^18.1.0",
```
[the remark this answers is not in the document]
Comment (caesarWHLee) [reply packages/draft-editor/package.json]: 感謝建議，已修正在[ba73ee4](https://github.com/mirror-media/Lilith/pull/213/commits/ba73ee4bf704294f61e236d0d55db1f347915b60)

File: packages/draft-editor/package.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "dev": "make dev",
+    "build": "make build",
+    "clean": "make clean",
+    "build-lib": "make build-lib"
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/readr-media/react.git",
```
[the remark this answers is not in the document]
Comment (caesarWHLee) [reply packages/draft-editor/package.json]: 感謝debug，已修正在[ba73ee4](https://github.com/mirror-media/Lilith/pull/213/commits/ba73ee4bf704294f61e236d0d55db1f347915b60)

**Comment by nickhsine** [comment]
LGTM 👍 👑 🥇

**Pull request merged**